url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/chunk/PNGFileChunk.java

```diff
@@ -0,0 +1,192 @@
+/**
+ * This file is distributed under a BSD-style license. See the included LICENSE.txt file
+ * for more information. 
+ * Copyright (c) 2009, James Halliday
+ * All rights reserved.
+ */
+package malictus.klang.chunk;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import malictus.klang.KlangConstants;
+import malictus.klang.KlangUtil;
+import malictus.klang.primitives.*;
+import malictus.klang.file.*;
+
+/**
+ * This class represents an entire PNG file (here called a 'chunk' for convenience).
+ * @author Jim Halliday
+ */
+public class PNGFileChunk extends ContainerChunk implements EditableContainerChunk {
+
+	public PNGFileChunk(KlangFile parentFile, long startpos, long endpos, long datastartpos, long dataendpos, int chunkType, String chunkName, ContainerChunk parentChunk, RandomAccessFile raf) throws IOException {
+		super(parentFile, startpos, endpos, datastartpos, dataendpos, chunkType, chunkName, parentChunk, raf);
+		raf.seek(datastartpos);
+		/*
+		 * ADD SUBCHUNKS
+		 */
+		raf.seek(datastartpos);
+		long pointer = raf.getFilePointer();
+		while (pointer < dataendpos) {
+    		Chunk cChunk = ChunkFactory.createChunk(parentFile, pointer, KlangConstants.CHUNKTYPE_PNG, this, raf, true, null, null);
+    		this.chunks.add(cChunk);
+    		pointer = cChunk.getEndPosition();
+    	}
+	}
+	
+	public void addChunkFromFile(File theFile, String name, int position, RandomAccessFile raff) throws IOException {
+		if (!(this.getParentFile() instanceof EditableFileBase)) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		long newsize = (this.endPosition - this.startPosition) + theFile.length();
+		if (newsize < 0) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		try {
+			//warn parent chunks (usually shouldn't be necessary, unless PNG chunk is embedded?)
+			this.chunkIsAboutToChangeSize(newsize, raff);
+			//now make the addition
+			long startingplace = 0;
+			if (position > this.getSubChunks().size() + 2) {
+				position = this.getSubChunks().size() + 1;
+			}
+			if (position == this.getSubChunks().size() + 1) {
+				if (this.getSubChunks().size() == 0) {
+					startingplace = this.endPosition;
+				} else {
+					startingplace = this.getSubChunks().get(position - 2).getEndPosition();
+				}
+			} else {
+				startingplace = this.getSubChunks().get(position - 1).getStartPosition();
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 54:62, 107:115.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.